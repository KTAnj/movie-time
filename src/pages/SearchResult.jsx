import React, { useCallback, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import Card from "../components/Card.jsx";
import { fetchMovies } from "../reducer/slices/movieSlice";
import InfiniteScroll from "react-infinite-scroll-component";
import { useSearchParams } from "react-router-dom";

const SearchResult = () => {
  const dispatch = useDispatch();
  const [searchParams] = useSearchParams();

  const query = searchParams.get("query");

  const {
    data: movies,
    page,
    hasNextPage,
    error,
    isLoading,
  } = useSelector((state) => state.movies);

  useEffect(() => {
    dispatch(fetchMovies({ page, query }));
  }, [query]);

  const fetchNextPage = useCallback(
    () => dispatch(fetchMovies({ page, query })),
    [page, query]
  );

  if (movies.length === 0 && isLoading) {
    return <h1>Loading ...</h1>;
  }

  if (movies.length === 0 && error) {
    return (
      <>
        <h2>Search results for &quot;{query}&quot;</h2>
        <h1 className="error">{error || "No data found"}</h1>
      </>
    );
  }

  return (
    <InfiniteScroll
      dataLength={movies.length || 0}
      next={fetchNextPage}
      loader={<h4>Loading ...</h4>}
      hasMore={hasNextPage}
    >
      <h2>Search results for &quot;{query}&quot;</h2>
      <div className="row">
        {movies &&
          movies.map((movie) => (
            <div className="col" key={`movie-${movie.imdbID}`}>
              <Card
                title={movie.Title}
                image={movie.Poster}
                id={movie.imdbID}
                year={movie.Year}
              />
            </div>
          ))}
      </div>
    </InfiniteScroll>
  );
};

export default SearchResult;
