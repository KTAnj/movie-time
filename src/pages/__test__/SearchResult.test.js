import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import SearchResult from "../SearchResult";
import * as ReactRouter from "react-router-dom";
import * as redux from "react-redux";

const mockDispatch = jest.fn();

const DUMMY_MOVIE = {
  imdbID: 1,
  Title: "Home Alone 2",
  Poster:
    "https://m.media-amazon.com/images/M/MV5BNDI1MzM0Y2YtYmIyMS00ODE3LTlhZjEtZTUyNmEzMTNhZWU5XkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg",
  Year: "2018",
};

const mockState = {
  query: "young",
  data: [DUMMY_MOVIE],
  error: null,
  isLoading: false,
};

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => mockDispatch,
  useSelector: () => mockState,
}));

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useSearchParams: () => [
    {
      get: () => "young",
    },
  ],
}));

function MockSearchResult() {
  return (
    <ReactRouter.BrowserRouter>
      <SearchResult />
    </ReactRouter.BrowserRouter>
  );
}

describe("should render Home component", () => {
  it("should match movie detail snapshot", () => {
    const tree = renderer.create(<MockSearchResult />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("render image correctly", () => {
    render(<MockSearchResult />);
    const imageElement = screen.getByAltText(DUMMY_MOVIE.Title);
    expect(imageElement).toBeVisible();
    expect(imageElement.getAttribute("src")).toContain(DUMMY_MOVIE.Poster);
  });

  it("render default image correctly", () => {
    const spy = jest.spyOn(redux, "useSelector");
    spy.mockReturnValue({ data: [{ ...DUMMY_MOVIE, Poster: "N/A" }] });
    render(<MockSearchResult />);
    const imageElement = screen.getByAltText(DUMMY_MOVIE.Title);
    expect(imageElement).toBeVisible();
    expect(imageElement.getAttribute("src")).toMatch(/default-movie.jpg$/);
  });

  it("render loading state correctly", async () => {
    const spy = jest.spyOn(redux, "useSelector");
    spy.mockReturnValue({ isLoading: true, data: [] });
    render(<MockSearchResult />);
    const loadingElement = await screen.findByText("Loading ...");
    expect(loadingElement).toBeVisible();
  });

  it("render error state correctly", async () => {
    const spy = jest.spyOn(redux, "useSelector");
    spy.mockReturnValue({ error: "Server Error", data: [] });
    render(<MockSearchResult />);
    const errorElement = await screen.findByText("Server Error");
    expect(errorElement).toBeVisible();
  });

  it("render no movie state correctly", async () => {
    const spy = jest.spyOn(redux, "useSelector");
    spy.mockReturnValue({ data: [] });
    render(<MockSearchResult />);
    const errorElement = await screen.findByText("No data found");
    expect(errorElement).toBeVisible();
  });
});
