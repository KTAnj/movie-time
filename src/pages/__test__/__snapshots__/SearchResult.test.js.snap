// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Home component shold match movie detail snapshot 1`] = `
<div
  className="infinite-scroll-component__outerdiv"
  style={{}}
>
  <div
    className="infinite-scroll-component "
    style={
      {
        "WebkitOverflowScrolling": "touch",
        "height": "auto",
        "overflow": "auto",
      }
    }
  >
    <h2>
      Search results for "
      young
      "
    </h2>
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div
          className="card"
        >
          <img
            alt="Home Alone 2"
            className="card-top-img"
            onClick={[Function]}
            src="https://m.media-amazon.com/images/M/MV5BNDI1MzM0Y2YtYmIyMS00ODE3LTlhZjEtZTUyNmEzMTNhZWU5XkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg"
          />
          <div
            className="card-title"
          >
            Home Alone 2
          </div>
          <small
            className="ml-auto"
          >
            2018
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render Home component should match movie detail snapshot 1`] = `
<div
  className="infinite-scroll-component__outerdiv"
  style={{}}
>
  <div
    className="infinite-scroll-component "
    style={
      {
        "WebkitOverflowScrolling": "touch",
        "height": "auto",
        "overflow": "auto",
      }
    }
  >
    <h2
      data-testid="searched-text"
    >
      Search results for "
      young
      "
    </h2>
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div
          className="card"
        >
          <img
            alt="Home Alone 2"
            className="card-top-img"
            onClick={[Function]}
            src="https://m.media-amazon.com/images/M/MV5BNDI1MzM0Y2YtYmIyMS00ODE3LTlhZjEtZTUyNmEzMTNhZWU5XkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg"
          />
          <div
            className="card-title"
          >
            Home Alone 2
          </div>
          <small
            className="ml-auto"
          >
            2018
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;
